docs/ReferenceGuides/UnderscoredAttributes.md

```diff
@@ -0,0 +1,428 @@
+changing the name mangling. This can be used to move declarations
+from a module to one of the modules it imports without breaking clients.
+
+Consider the following example where a framework ToasterKit needs
+to move some APIs to a lower-level framework ToasterKitCore.
+Here are the necessary changes:
+
+1. Add a linker flag `-reexport_framework ToasterKitCore` for ToasterKit.
+   This ensures all symbols defined in ToasterKitCore will be accessible during
+   runtime via ToasterKit, so existing apps continue to run.
+2. In ToasterKit, use `@_exported import ToasterKitCore`.
+   This ensures existing source code that only imports ToasterKit continues to
+   type-check.
+3. Move the necessary declarations from ToasterKit to ToasterKitCore.
+   The moved declaration should have two attributes:
+   - `@available` indicating when the declaration was introduced in ToasterKit.
+   - `@_originallyDefinedIn` indicating the original module and when the
+     declaration was moved to ToasterKitCore.
+   ```swift
+   @available(toasterOS 42, *)
+   @_originallyDefinedIn(module: "ToasterKit", toasterOS 57)
+   enum Toast {
+   case underdone
+   case perfect
+   case burnt
+   }
+   ```
+4. Add Swift compiler flags `-Xfrontend -emit-ldadd-cfile-path -Xfrontend /tmp/t.c`
+   to ToasterKitCore's build settings. Add the emitted `/tmp/t.c` file to
+   ToasterKit's compilation.
+   This ensures when an app is built for deployment targets prior to the symbols' move,
+   the app will look for these symbols in ToasterKit instead of ToasterKitCore.
+
+More generally, mutliple availabilities can be specified, like so:
+
+```swift
+@available(toasterOS 42, bowlOS 54, mugOS 54, *)
+@_originallyDefinedIn(module: "ToasterKit", toasterOS 57, bowlOS 69, mugOS 69)
+enum Toast { ... }
+```
+
+## `@_private(sourceFile: "FileName.swift")`
+
+Fully bypasses access control, allowing access to private declarations
+in the imported module. The imported module needs to be compiled with
+`-Xfrontend -enable-private-imports` for this to work.
+
+## `@_semantics("uniquely.recognized.id")`
+
+Allows the optimizer to make use of some key invariants in performance critical
+data types, especially `Array`. Since the implementation of these data types
+is written in Swift using unsafe APIs, without these attributes the optimizer
+would need to make conservative assumptions.
+
+Changing the implementation in a way that violates the optimizer's assumptions
+about the semantics results in undefined behavior.
+
+## `@_show_in_interface`
+
+## `@_silgen_name("cName")`
+
+Changes the symbol name for a function, similar to an ASM label in C,
+except that the platform symbol mangling (leading underscore on Darwin)
+is maintained.
+
+Since this has label-like behavior, it may not correspond to any declaration;
+if so, it is assumed that the function is implemented in C.
+
+A function defined by `@_silgen_name` is assumed to use the Swift ABI.
```

Comment:
I believe there's another document explaining the different uses of `@_silgen_name` in more detail. Perhaps it could be linked to here?

Reply:
Add link to Standard Library Programmer's Manual.